clients/gax/lib/google_api/gax/connection.ex

```diff
@@ -192,6 +205,20 @@ defmodule GoogleApi.Gax.Connection do
     Keyword.put(output, :body, body)
   end
 
+  defp extract_metadata(body_params) do
+    case Keyword.fetch(body_params, :metadata) do
+      :error -> {nil, body_params}
+      {:ok, meta} -> {meta, Keyword.delete(body_params, :metadata)}
+    end
+  end
+
+  defp try_encode_multipart_field(data) do
+    case Poison.encode(data) do
+      {:ok, json} -> {json, "application/json"}
+      _ -> {data, "application/octet-stream"}
+    end
```

Comment:
There is an ambiguity for types (like text strings) that can be legally JSON-encoded but could also be raw iodata (for example, the string `hello` which Poison is happy to encode as `"hello"`). So the current PR still prevents us from uploading plain text data because Poison would accept it, and thus would mangle it by surrounding it with quotes (as well as possibly escaping characters).

There is currently no use case across any of the Google REST APIs for multipart-body requests other than file upload. Therefore, in ambiguous cases, we should be safe assuming that the input is intended as iodata. The only time I think we should json-encode an input is if it is a data structure (such as the various data types defined by APIs). So what I suggest is, change the logic here so it Poison encodes the data only if it pattern matches against a map (which will catch all maps and structs, i.e. all API-defined data structures). Otherwise, just pass it through as octet-stream.

Reply:
I agree. To add, we should extract `:contentType` from `metadata` and fallback on `application/octet-stream`.

```elixir
headers: [{:"Content-Type", Map.get(meta, :contentType, "application/octet-stream")}]
```

Reply:
I wouldn't be opposed to setting the Content-Type header, but is probably not as important, because the backend service doesn't actually use it (since it picks up the mime type from the metadata). The important change is not to Poison encode unless we have to (because the input is a data structure rather than iodata).

Reply:
If I understood this error correctly https://github.com/googleapis/elixir-google-api/pull/3638#issuecomment-638880383, the request fails if the metadata `contentType` does not match the multipart `Content-Type` header.

Reply:
Huh. That error is surprising to me; that's not how I thought the upload backend worked. But if we're getting that, then, yeah, we should set the Content-Type also.

Reply:
I just want to make sure I've got the flow correct in my head, so correct me if I'm mistaken on any part:

1. If `data` is a map, try encoding it as JSON.
2. If `data` is not a map, or fails to encode to a JSON payload properly, we return the data as-is (well, as a binary).
3. If `meta` is a map, we try to extract the payload content type from it, falling back to `application/octet-stream`.
4. If `meta` is not a map, we default to `application/octet-stream` for the content type.

I'm good with this. My only _possible_ concern is whether there are any situations where you might have more than one payload in a request. In such a case, the content type in the metadata would try to apply to all parts of the body, which might not be correct. But I can't think of, nor can I see with a cursory check, any functions that might try having multiple "body" parts that would use separate content types.

Reply:
Yeah, that's about right. The tweak I'd make is that 2. if `data` is a map but fails to JSON encode, then it is converted to a binary (using `Kernel.inspect/1`), otherwise if it is not a map, we assume it is a properly-formed `iodata` and pass it through as-is. At least, that's the implementation I pushed to your branch. I'm actually not sure if we should catch JSON encoding errors. Maybe we should just let those crash on the assumption that the only valid inputs should be proper `iodata` or JSON-encodable maps.

Reply:
Yeah, as I think about this, I don't see that there would be a reasonable use case for passing an Elixir map to upload and expecting that it would be encoded according to its `inspect/1` output. I'm going to change the implementation to let poison crash in such a case. I think that's better than silently uploading something we don't expect. If someone actually does intentionally want the `inspect/1` output uploaded, they can call it themselves and pass in the resulting binary.

Reply:
@kolorahl @jayjun I pushed a change to the PR branch according to the above discussion, along with tests. Let me know if either of you have further comments. Otherwise, I'll merge this and push a release on Monday.

Reply:
It's funny you mentioned that. I woke up this morning thinking the exact same thing: _when would you have a map that you **don't** want converted to a JSON payload?_

I definitely like this implementation better. If it's a map and can't be converted to JSON, fail. Otherwise accept it as `iodata` and use `meta` to possibly determine the content type. 👍

Reply:
Looks good to me. `Poison.encode!` is a good call.